File: acapella_api/app/Models/Subscription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

class Subscription extends Model
{
    use SoftDeletes;

    protected $guarded = ['id'];

    protected $fillable = [
        'user_id',
        'plan',
        'amount',
        'currency',
        'status',
        'idempotency_key',
        'mongike_payment_id',
        'started_at',
        'expires_at',
        'grace_period_ends_at',
    ];

    protected $appends = ['formatted_amount'];

    protected function casts(): array
    {
        return [
            'amount'               => 'integer',
            'started_at'           => 'datetime',
            'expires_at'           => 'datetime',
            'grace_period_ends_at' => 'datetime',
        ];
    }

    // ── Accessors ─────────────────────────────────────────────────────────────

    /**
     * Formatted TZS price as displayed in the UI, e.g. "TZS 5,000/="
     */
    public function getFormattedAmountAttribute(): string
    {
        return 'TZS ' . number_format($this->amount) . '/=';
    }

    // ── Business Logic ────────────────────────────────────────────────────────

    public function isActive(): bool
    {
        // Subscription is active if status is active and either:
        // 1. No expiration date, or
        // 2. Not yet expired, or
        // 3. Expired but still within grace period
        return $this->status === 'active'
            && ($this->expires_at === null || $this->expires_at->isFuture() || $this->isInGracePeriod());
    }

    public function isInGracePeriod(): bool
    {
        return $this->grace_period_ends_at !== null
            && $this->grace_period_ends_at->isFuture();
    }

    // ── Relationships ─────────────────────────────────────────────────────────

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    // ── Scopes ───────────────────────────────────────────────────────────────

    public function scopeWithHistory($query)
    {
        return $query->withTrashed();
    }
}

File: acapella_app/lib/models/subscription.dart
class Subscription {
  final int id;
  final int userId;
  final String plan;
  final double amount;
  final String formattedAmount; // Formatted TZS representation from backend
  final String currency;
  final String status;
  final String idempotencyKey;
  final String? snippePaymentId;
  final DateTime? startedAt;
  final DateTime? expiresAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Subscription({
    required this.id,
    required this.userId,
    required this.plan,
    required this.amount,
    required this.formattedAmount,
    this.currency = 'TZS',
    required this.status,
    required this.idempotencyKey,
    this.snippePaymentId,
    this.startedAt,
    this.expiresAt,
    this.createdAt,
    this.updatedAt,
  });

  bool get isActive => status == 'active' && (expiresAt == null || expiresAt!.isAfter(DateTime.now()));

  factory Subscription.fromJson(Map<String, dynamic> json) {
    return Subscription(
      id: json['id'],
      userId: json['user_id'],
      plan: json['plan'],
      amount: (json['amount'] ?? 0).toDouble(),
      formattedAmount: json['formatted_amount'] ?? '',
      currency: json['currency'] ?? 'TZS',
      status: json['status'],
      idempotencyKey: json['idempotency_key'],
      snippePaymentId: json['mongike_payment_id'],
      startedAt: json['started_at'] != null ? DateTime.parse(json['started_at']) : null,
      expiresAt: json['expires_at'] != null ? DateTime.parse(json['expires_at']) : null,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'plan': plan,
      'amount': amount,
      'currency': currency,
      'status': status,
      'idempotency_key': idempotencyKey,
      'mongike_payment_id': snippePaymentId,
      'started_at': startedAt?.toIso8601String(),
      'expires_at': expiresAt?.toIso8601String(),
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}
